src/processor/operator/copy/copy_node.cpp

```diff
@@ -82,16 +82,25 @@ void CopyNode::executeInternal(ExecutionContext* context) {
     sharedState->logCopyNodeWALRecord(copyNodeInfo.wal);
     while (children[0]->getNextTuple(context)) {
         // All tuples in the resultSet are in the same data chunk.
-        auto numTuplesToAppend = ArrowColumnVector::getArrowColumn(
+        auto numTuplesToAppend = ArrowColumnVector::getChunkedArrowColumn(
             resultSet->getValueVector(copyNodeInfo.dataColumnPoses[0]).get())
                                      ->length();
+        auto nodeOffset = nodeOffsetVector->getValue<int64_t>(
+                              nodeOffsetVector->state->selVector->selectedPositions[0]) -
+                          1;
         uint64_t numAppendedTuples = 0;
         while (numAppendedTuples < numTuplesToAppend) {
             auto numAppendedTuplesInNodeGroup = localNodeGroup->append(
                 resultSet, copyNodeInfo.dataColumnPoses, numTuplesToAppend - numAppendedTuples);
             numAppendedTuples += numAppendedTuplesInNodeGroup;
             if (localNodeGroup->isFull()) {
-                auto nodeGroupIdx = sharedState->getNextNodeGroupIdx();
+                node_group_idx_t nodeGroupIdx;
+                if (copyNodeInfo.preservingOrder) {
+                    nodeGroupIdx = nodeOffset / StorageConstants::NODE_GROUP_SIZE - 1;
```

Comment:
As we change the `nodeOffset` to track the start nodeOffset, instead of the end one, we should also change L99 here. You should be able to directly reuse `getNodeGroupIdxFromNodeOffset` function.